include/nccl_ofi.h

```diff
@@ -475,24 +493,38 @@ struct nccl_net_ofi_ep {
 	 * endpoint if reference counter becomes zero. Must be
 	 * protected by lock stored in base_dev.
 	 *
-	 * @param	ep
-	 * 		The endpoint (itself) to be released.
 	 * @param 	skip_lock
 	 * 		false, taking domain lock by default.
 	 * 		ture, not taking domain lock when caller takes it.
 	 * @param	force_cleanup
 	 * 		false, not release when endpoint has ref count.
 	 * 		true, release no matter endpoint has ref count or not.
 	 */
-	int (*release_ep)(nccl_net_ofi_ep_t *ep, bool skip_lock, bool force_cleanup);
+	virtual int release_ep(bool skip_lock, bool force_cleanup);
 
-/* private */
-	/* pure virtual function called when resources associated with
-	 * the ep should be destroyed.  Device lock will be held when
-	 * this function is called.
+	/**
+	 * @brief	Gets the base endpoint reference count.
+	 */
+	inline int get_ref_cnt() {
```

Comment:
Pretty sure we shouldn't have a get_ref_cnt(), and the increment / decrement should be sufficient.  I'm also not entirely sure i love this pattern of reference management (this is probably a place for a smart pointer), but I'll accept that's probably later work.

Reply:
Good point for not needing the `get_ref_cnt` function. Currently the only time the plugin directly accesses `ref_cnt` value is in base and rdma endpoint definitions of `release_ep` to check if its equal to 0. I could either directly use `this->ref_cnt` if I make `ref_cnt` protected instead of private, or just add a separate inline function that returns `ref_cnt == 0`.

I'll take a quick look through smart pointers (`std::shared_pointer` and `std::enable_shared_from_this`) to see if I could integrate it without too much effort, if it takes me a while to get the hang of using them I might punt it until a later PR.

Reply:
punt it to later.  You need everything that takes a reference to be c++ized, which is not a state we have right now.